app.py

```diff
@@ -297,6 +326,32 @@ def parse():
     # reason.
     lang = body.get('lang', requested_lang())
 
+# begin
+    ad_name = body['adventure_name']
+    cur_level = int(body['level'])
+    ad_index = ad_name2index[(cur_level,ad_name)]
+
+    ad_index_db = DATABASE.get_ad(current_user () ['username'])
+    level_db = DATABASE.get_level(current_user () ['username'])
+
+    level_def = load_yaml_rt(f'coursedata/level-defaults/{lang}.yaml')
+    commands = level_def[cur_level]['commands']
+
+    if level_db == cur_level:
+        # may unlock, flag may be set to 1
+        index = 0
+        for com in commands:
+            demo_code = com['demo_code']
+            demo_code += '\n'
+            if code == demo_code:
+                RUN_COMMAND[index] = 1
+            index += 1
+        flag = 1
+        for i in RUN_COMMAND:
+            if i == 0:
+                flag = 0
```

Comment:
It is unclear to me what this does, can you add a comment or make a function?

Reply:
After the users have tried all the commands, they can unlock the next level. I record the infomation in the RUN_COMMAND array. If the user run a command, the  RUN_COMMAND[index] will be set to 1. If all the members of RUN_COMMAND array are set to 1, the next level will be unlocked.

Reply:
Can we add this as a comment in the source? Combined with better variable names it might become understandable to me.